### DISCUSSION TITLE
I'm creating a Fish linter (fishcheck) and could use your input!

### DISCUSSION THREAD

**Discussion by mattmc3** [fish-shell/fish-shell#10725] [General]
There's [been discussion](https://stackoverflow.com/questions/42660371/is-there-a-linter-for-fish-like-there-is-for-bash-with-shellcheck) about creating [a Fish linter](https://github.com/fish-shell/fish-shell/issues/8357) before. Fish largely doesn't need much in the way of linting compared to Bash, but there are some runtime errors that `fish --no-execute` won't catch, and some stylistic issues that `fish_indent` won't correct.

For example, I have seen new users do things like:

```fish
# This isn't a runtime error, but it's very, very wrong
if count $args -gt 0
    echo "foo"
end
```

when they meant to do:

```fish
if test (count $args) -gt 0
    echo "foo"
end
```

I admit that I still frequently get bit by accidentally shadowing a local:

```fish
set -l foo bar
if true
  set -l foo baz  # -l here is wrong
end
echo $foo  # oops, it's bar not baz
```

So far, I have a [few checks implemented and working here](https://github.com/mattmc3/fishcheck/wiki), and I could use community input. Are there any basic mistakes you'd like a linter to catch?

NOTE: I have no intention of making this project a full blown language parser/LSP. Poking around in `read --tokenize` is as far as I'm looking to go at the moment.

Here's output from my pre-alpha crack at fishcheck:
<img width="644" alt="Screenshot 2024-09-16 at 3 55 08 PM" src="https://github.com/user-attachments/assets/857f263a-562b-4d1c-a733-33967fd0984b">

I'm already looking at [shellcheck](https://gist.github.com/nicerobot/53cee11ee0abbdc997661e65b348f375#file-_shellcheck-md) to see what's worth borrowing as well. Here is [my running list of checks](https://github.com/mattmc3/fishcheck/blob/main/checks.md).

**Comment by ndonfris** [comment]
Hey, don't know if you've seen the project I'm working on, [fish-lsp](https://github.com/ndonfris/fish-lsp), but I thought it's worth mentioning since it's got a decent amount of linting support, currently and has much more to come in the future releases.

I have an open discussion as well about basic features supported [here](https://github.com/ndonfris/fish-lsp/discussions/37), and am currently adding support for the following features:

- unused variables + unreachable code blocks 
- automating `argparse` completions for autoloaded functions, and notifying unused `argparse` options
- providing code actions to auto-fix issues
- many other features (in up-coming versions). 

Many of the features you currently support overlap with the currently supported diagnostic error numbers I am providing in the fish-lsp, so I thought it'd be helpful to work together on this.

I am looking for support building the fish-lsp, if you're interested in working on this aspect of the project. Good job on your work so far!

**Reply by mattmc3** [reply]
This is incredible! I'll have a look and see how we can collaborate. I have a running list of about 30 mistakes Fish users/n00bs make, most of them around `test` and bash-isms. I'll polish that up and send you what I have so far. I still think there's room for a simple CLI linter like shellcheck for Fish, but tying into a fish-lsp backend isn't out of the question.

**Reply by edouard-lopez** [reply]
That's awesome! :clap: 
Thank you so much for the VSCode extension :heart_eyes:! That will help me a lot for [pure](https://github.com/pure-fish/pure) development.
